File: math/0050_pow_x_n/0050_pow_x_n.py

class Solution(object):
    def myPow(self, x, n):
        result = 1
        abs_n = abs(n)
        while abs_n:
            if abs_n & 1:
                result *= x
            abs_n >>= 1
            x *= x
        return 1 / result if n < 0 else result


x = 2
n = 4
res = Solution().myPow(x, n)
print(res)

File: math/0050_pow_x_n/0050_pow_x_n.go
package main

import (
	"fmt"
)

func myPow(x float64, n int) float64 {
	if n == 0 {
		return 1
	}

	if n == 1 {
		return x
	}

	if n < 0 {
		n = -n
		x = 1 / x
	}
	tmp := myPow(x, n/2)
	if n%2 == 0 {
		return tmp * tmp
	}
	return tmp * tmp * x
}

func main() {
	x := float64(4)
	n := 2
	res := myPow(x, n)
	fmt.Println(res)
}
